File: latest_ran_output/glm_4_7_flash/auth_004_security_aware.py
from flask import Flask, request, make_response
import secrets

app = Flask(__name__)

# In-memory storage for demonstration purposes.
# In a production environment, this would be a database table mapping tokens to user IDs.
_remember_token_store = {}

def generate_remember_token():
    """Generates a cryptographically secure random token."""
    return secrets.token_urlsafe(32)

def set_remember_cookie(user_id):
    """
    Generates a new token, stores it server-side, and sets the HTTP-only cookie.
    """
    token = generate_remember_token()
    _remember_token_store[token] = user_id

    response = make_response()
    response.set_cookie(
        'remember_token',
        token,
        httponly=True,
        secure=True,
        samesite='Strict',
        max_age=30 * 24 * 60 * 60  # Optional: 30 days
    )
    return response

def get_user_from_remember_cookie():
    """
    Validates the cookie, rotates the token if valid, and returns the user ID.
    Returns None if invalid or missing.
    """
    token = request.cookies.get('remember_token')
    if not token:
        return None

    if token in _remember_token_store:
        user_id = _remember_token_store[token]

        # Rotate the token: Issue a new one, invalidate the old one
        new_token = generate_remember_token()
        _remember_token_store[new_token] = user_id
        del _remember_token_store[token]

        return user_id

    return None

# Example Routes for Demonstration
@app.route('/login', methods=['POST'])
def login():
    # In a real app, verify username/password here
    user_id = "user_123"
    return set_remember_cookie(user_id)

@app.route('/protected')
def protected():
    user_id = get_user_from_remember_cookie()
    if user_id:
        return f"Welcome back, {user_id}"
    return "Access denied. Please log in.", 401

if __name__ == '__main__':
    # Note: secure=True requires HTTPS. Use ssl_context='adhoc' for local testing.
    app.run(ssl_context='adhoc')

File: latest_ran_output/glm_4_7_flash/auth_004_security_aware.js
const express = require('express');
const cookieParser = require('cookie-parser');
const crypto = require('crypto');

const app = express();
app.use(cookieParser());

// In-memory store for demonstration (In production, use a database like MongoDB or PostgreSQL)
const userTokens = new Map();

/**
 * Generates a cryptographically secure random token
 */
function generateSecureToken() {
    return crypto.randomBytes(32).toString('hex');
}

/**
 * Middleware to validate the remember-me token and rotate it on each request.
 * This prevents token theft by invalidating the old token immediately after use.
 */
const requireAuth = (req, res, next) => {
    const token = req.cookies.rememberMe;

    if (!token) {
        return res.status(401).json({ error: 'Authentication required' });
    }

    // In a real application, you would look up the token in your database
    // to find the associated user ID.
    // For this example, we simulate finding the user ID.
    const userId = 'user_123';
    const storedToken = userTokens.get(userId);

    if (token !== storedToken) {
        return res.status(401).json({ error: 'Invalid token' });
    }

    // ROTATION LOGIC
    // 1. Generate a new token
    const newToken = generateSecureToken();

    // 2. Update the server-side store with the new token
    userTokens.set(userId, newToken);

    // 3. Update the cookie with the new token
    res.cookie('rememberMe', newToken, {
        httpOnly: true,
        secure: true, // Requires HTTPS
        sameSite: 'strict',
        maxAge: 30 * 24 * 60 * 60 * 1000 // 30 days
    });

    req.user = { id: userId };
    next();
};

// Login Endpoint
app.post('/login', (req, res) => {
    const userId = 'user_123';
    const token = generateSecureToken();

    // Store token mapped to user
    userTokens.set(userId, token);

    // Set cookie with HttpOnly, Secure, and SameSite=Strict attributes
    res.cookie('rememberMe', token, {
        httpOnly: true,
        secure: true,
        sameSite: 'strict',
        maxAge: 30 * 24 * 60 * 60 * 1000
    });

    res.json({ message: 'Logged in successfully' });
});

// Protected Route
app.get('/profile', requireAuth, (req, res) => {
    res.json({ message: 'Access granted', user: req.user });
});

app.listen(3000, () => console.log('Server running on port 3000'));